File: travel_blog/schema/mutations/articles.py
from graphene import Field, Int, Mutation, String

from ..objects import Article
from ...models import (
    db,
    Article as ArticleModel,
)
from ...tools import normalize_id

class CreateArticle(Mutation):
    class Input:
        section_id = Int()
        name = String()

    article = Field(lambda: Article)

    @staticmethod
    def mutate(root, args, context, info):
        article = ArticleModel(
            str_id=normalize_id(args['name']),
            **args
        )
        db.session.add(article)
        db.session.commit()
        return CreateArticle(article=None)

class DeleteArticle(Mutation):
    class Input:
        id = Int()

    article = Field(lambda: Article)

    @staticmethod
    def mutate(root, args, context, info):
        id = int(args['id'])
        article = Article.get_query(context).get_or_404(id)
        article.deleted = True
        db.session.commit()
        return DeleteArticle(article=article)

class DestroyArticle(Mutation):
    class Input:
        id = Int()

    article = Field(lambda: Article)

    @staticmethod
    def mutate(root, args, context, info):
        id = int(args['id'])
        article = Article.get_query(context).get_or_404(id)
        db.session.delete(article)
        db.session.commit()
        return DestroyArticle(article=None)

class UndeleteArticle(Mutation):
    class Input:
        id = Int()

    article = Field(lambda: Article)

    @staticmethod
    def mutate(root, args, context, info):
        id = int(args['id'])
        article = Article.get_query(context).get_or_404(id)
        article.deleted = False
        db.session.commit()
        return UndeleteArticle(article=article)

File: src/mutations/articles.js
import { graphql, gql } from 'react-apollo';

import * as queries from '../queries';

export const createArticle = graphql(gql`
  mutation($sectionId: Int!, $name: String!) {
    createArticle(sectionId: $sectionId, name: $name) {
      article {
        id
        strId
        name
        deleted
      }
    }
  }
`, {
  name: 'createArticle',
  options: props => ({
    refetchQueries: [
      { query: queries.section, variables: { strId: props.sectionId } },
    ],
  }),
});

export const deleteArticle = graphql(gql`
  mutation($id: Int!) {
    deleteArticle(id: $id) {
      article {
        id
        deleted
      }
    }
  }
`, {
  name: 'deleteArticle',
  options: props => ({
    refetchQueries: [
      { query: queries.section, variables: { strId: props.sectionId } },
    ],
  }),
});

export const destroyArticle = graphql(gql`
  mutation($id: Int!) {
    destroyArticle(id: $id) {
      article {
        id
      }
    }
  }
`, {
  name: 'destroyArticle',
  options: props => ({
    refetchQueries: [
      { query: queries.section, variables: { strId: props.sectionId } },
    ],
  }),
});

export const undeleteArticle = graphql(gql`
  mutation($id: Int!) {
    undeleteArticle(id: $id) {
      article {
        id
        deleted
      }
    }
  }
`, {
  name: 'undeleteArticle',
  options: props => ({
    refetchQueries: [
      { query: queries.section, variables: { strId: props.sectionId } },
    ],
  }),
});
